### DISCUSSION TITLE
How to get pango to find Windows  fonts

### DISCUSSION THREAD

**Discussion by bangusi** [ImageMagick/ImageMagick#2644] [Help]
I am on Windows. The following command 
`magick convert -size 1440x900 -background black  -font "c:\windows\fonts\Times.ttf" -fill white -gravity center   pango:"<big><big>Large</big></big> <!-- --> <small>Small</small>" pango_formatting.png`
 
gives the output
`** (process:26204): WARNING **: 07:46:45.625: couldn't load font "c:\windows\fonts\Times.ttf Not-Rotated 17.279296875", falling back to "Sans Not-Rotated 17.279296875", expect ugly output.`

Other protocols like label and caption have no trouble finding the font.

**Comment by snibgo** [comment]
As far as I know, "-font path\filename.ttf" isn't recognised by pango.

Try these two methods:

1. Use "-font ABCD" before the pango, where ABCD is a font that is listed in "-list font".

2. In your pango, include "<span font="ABCD">Your text here</span>".

**Comment by bangusi** [comment]
This does not work either

`magick convert -size 1440x900 -background black  -font "Times" -fill white -gravity center pango:"<big><big>Large</big></big> <!-- --> <small>Small</small>" pango_formatting.png`

I am not sure what you mean by a " font that is listed in "-list font". "

**Comment by snibgo** [comment]
For v7, I suggest you use "magick", not "magick convert".

"magick -list font" will show you what fonts IM can see on your computer. For example:

```
magick -list font |grep imes

  Font: Times-New-Roman
    family: Times New Roman
    glyphs: c:\windows\fonts\times.ttf
  Font: Times-New-Roman-Bold
    family: Times New Roman
    glyphs: c:\windows\fonts\timesbd.ttf
  Font: Times-New-Roman-Bold-Italic
    family: Times New Roman
    glyphs: c:\windows\fonts\timesbi.ttf
  Font: Times-New-Roman-Italic
    family: Times New Roman
    glyphs: c:\windows\fonts\timesi.ttf
```
You will see that on my computer, IM can see no font called "Times".

**Comment by bangusi** [comment]
```
>magick -list font |grep imes
  Font: Times-New-Roman
    family: Times New Roman
    glyphs: c:\windows\fonts\times.ttf
  Font: Times-New-Roman-Bold
    family: Times New Roman
    glyphs: c:\windows\fonts\timesbd.ttf
  Font: Times-New-Roman-Bold-Italic
    family: Times New Roman
    glyphs: c:\windows\fonts\timesbi.ttf
  Font: Times-New-Roman-Italic
    family: Times New Roman
    glyphs: c:\windows\fonts\timesi.ttf
```
	
Still this command does not work
`magick -size 1440x900 -background black -font Times-New-Roman-Italic -fill white -gravity center pango:"<big><big>Large</big></big> <!-- --> <small>Small</small>" pango_formatting.png`
But this one works
`magick -size 1440x900 -background black -font Times-New-Roman-Italic -fill white -gravity center label:"<big><big>Large</big></big> <!-- --> <small>Small</small>" pango_formatting.png`

Does it ( the pango ) work on your machine?

**Comment by fmw42** [comment]
Do you have Cairo/Pango installed as a delegate in Imagemagick

**Comment by bangusi** [comment]
@fmw42 
I see these folders:
```
D:\ImageMagick-Windows\cairo\
D:\ImageMagick-Windows\pango\
```
Does it mean I have these installed as 'delegates'
I don't know.. sorry I am very new to ImageMagick. Discovered it just over a month ago.

BTW I compiled the ImageMagick from sources following the instructions.
I selected a static CRT build.

**Reply by fmw42** [reply]
What do you get from 

magick -version

does it list Cairo or Pango?

Add -verbose to your command and let us know what you get returned.

magick -verbose -size 1440x900 -background black -font Times-New-Roman-Italic -fill white -gravity center pango:"<big><big>Large</big></big>  <small>Small</small>" pango_formatting.png

**Reply by fmw42** [reply]
You may have to put the font into the markup language. See the documents listed at https://imagemagick.org/Usage/text/#pango

**Comment by bangusi** [comment]
magick -version
Version: ImageMagick 7.0.10-31 Q16 x64 2020-09-25 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 191627043
Features: Cipher DPC HDRI OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib

`magick -verbose -size 1440x900 -background black -font Times-New-Roman-Italic -fill white -gravity center pango:"Large Small" pango_formatting.png`

** (process:23172): WARNING **: 14:04:30.788: couldn't load font "Times-New-Roman-Italic Not-Rotated 12", falling back to "Sans Not-Rotated 12", expect ugly output.
pango:Large Small=>Large Small PANGO 1440x900 1440x900+0+0 16-bit sRGB 0.063u 0:00.056
pango:Large Small=>pango_formatting.png PANGO 1440x900 1440x900+0+0 16-bit sRGB 0.438u 0:00.105

**Reply by fmw42** [reply]
I think you can only use Cairo/Pango approved fonts and they should be used inside the PANGO markup. I have always had trouble specifying fonts for Pango.

**Comment by snibgo** [comment]
Hmm. With IM v7/0/8-64, it works fine when the font name is a single word:
```
magick -size 350x200 -font Sylfaen pango:"hello world" p.png
```
When the font name is two words, we need a space. (Put this another way, we use the font family instead of the font name.)
```
magick -size 350x200 -font "Yu Gothic" pango:"hello world" p.png
```
But this doesn't work when the font name has three words:
```
magick -size 350x200 -font "Times New Roman" pango:"hello world" p.png

** (magick.exe:8468): WARNING **: couldn't load font "Times New Not-Rotated 12",
 falling back to "Sans Not-Rotated 12", expect ugly output.
```
Note that the error message has picked up just the first two words, "Times New"

I suspect there is a bug.

**Reply by fmw42** [reply]
Makes sense about the font family. From what I could tell in the past, one puts the font family in Pango markup and that worked.
